src/app/components/datacard/datacard.component.ts

```diff
@@ -1,57 +1,116 @@
+export class DatacardComponent implements OnInit, OnDestroy {
+    private suscription: Subscription
+    countries: CountryApi[]
+    countryReport: CountryReport
+
     confirmedCases: any
+    iniciar: boolean
+    showCountry: boolean
     loading: boolean
     error: boolean
-    warning: boolean = false
+    warning: boolean
     mensajeError: string
-    iniciar: boolean
+    summary: any[] = []
+
+    constructor(private covid: CovidService, private router: ActivatedRoute, private adapter: CountryReportAdapter) {}
+
+    ngOnInit(): void {
+        this.stateHandler("init")
+        this.suscription = this.covid.getCountriesSummary().subscribe((countries: CountryApi[]) => {
+            this.countries = countries
 
-    get showCountry(): string {
-        return this._showCountry
+            this.router.params.subscribe((params) => {
+                if (Object.keys(params).length !== 0) {
+                    this.confirmedCases = this.getCountrySummary(params.country)
+                    this.stateHandler("data")
+                } else if (Object.keys(params).length === 0) {
+                    this.stateHandler("init")
+                }
+            })
+        })
     }
 
-    @Input()
-    set showCountry(value: string) {
-        this._showCountry = value
-        this.loading = true
-        this.error = false
-        this.warning = false
-        this.iniciar = false
-        if (value) {
-            this.covid
-                .getCountryConfirmed(this.showCountry) //
-                .subscribe(
-                    (data) => {
-                        this.confirmedCases = data
-                        this.loading = false
-                        if (data === undefined) {
-                            this.mensajeError = "There is no data for this country"
-                            this.warning = true
-                        }
-                    },
-                    (errorServicio) => {
-                        this.error = true
-                        this.loading = false
-                        this.mensajeError = "No server connection"
-                    }
-                )
-        } else {
-            this.mensajeError = "Select a country to see COVID-19 confirmed cases"
-            this.loading = false
-            this.iniciar = true
+    ngOnDestroy() {
+        this.suscription.unsubscribe()
+    }
+
+    // getConfirmed(country): object {
+    //     return this.summary.filter((pais) => pais.Slug === country).map((item) => ({ country: item.Country, confirmed: item.TotalConfirmed, date: item.Date }))[0]
+    // }
+
+    getCountrySummary(countryName: string): CountryReport {
+        this.stateHandler("loading")
+        if (this.countries) {
+            const foundCountry = this.countries.find((country) => country.Slug === countryName)
+            return this.adapter.adapt(foundCountry)
         }
     }
 
-    constructor(private covid: CovidService) {}
+    stateHandler(state: string) {
```

Comment:
This block of code has different responsibilities, try to separate the code in functions with a single responsibility.

Reply:
I divided the stateHandler into four different methods which I'll call separately where needed.